src/Lookup.hs

```diff
@@ -28,6 +29,27 @@ lookupInEnv path@(SymPath (p : ps) name) env =
         Just parent -> lookupInEnv path parent
         Nothing -> Nothing
 
+lookupInterfaceOrType :: Context -> SymPath -> Maybe Binder
```

Comment:
Generally speaking, I think we should move toward defining lookups against `Context` instead of `Env`.

There's tons of instances, currently, of 

```haskell
let global = contextGlobalEnv ctx
      typeE = contextTypeEnv ctx
```

etc., but we almost *never* (if ever?) manipulate the `env` before performing a lookup. Almost every function that performs a lookup just grabs an env as is to pass to a lookup function.

Instead, I think we should handle such env delving internally in `Lookup` and simply indicate what we're doing by well-named functions, the details of which remain internal:

`lookupGlobalBinding :: Context -> ...`
`lookupInternalBinding :: Context -> ...`
`lookupType :: Context -> ...`

This would enable us to remove all those noisy local bindings to envs. WDYT? @carp-lang/maintainers

Reply:
We can also perform further filtering internally. e.g. the function here, `lookupInterfaceOrType` could also have a specific version `lookupInterface` that peek into the XObj and only return it if it is in fact an interface.

```
lookupInterface :: Context -> Maybe Binder
lookupInterface ctx path = 
  case lookupInterfaceOrType ctx path of
      Just binder@(Binder _ <InterfaceXObj>) -> Just binder
      _ -> Nothing
```

Defining such functions in Lookup would eliminate tons of inline matches on Binder XObj contents in other functions as well.

Reply:
Basically, it'd make us be able to leverage `Maybe`, its monad, functor, applicative, etc. as an interface/context for our computations in more places and eliminate tons of `lets` and `case`s

Reply:
I'd say yes, yes and yes.

Reply:
Sounds great!